std/file.d

```diff
@@ -353,22 +377,36 @@ version (Posix) private void[] readImpl(const(char)[] name, const(FSChar)* namez
     stat_t statbuf = void;
     cenforce(fstat(fd, &statbuf) == 0, name, namez);
 
+    if (offset >= statbuf.st_size)
+        offset = to!size_t(statbuf.st_size + 1);
     immutable initialAlloc = min(upTo, to!size_t(statbuf.st_size
-        ? min(statbuf.st_size + 1, maxInitialAlloc)
+        ? min(statbuf.st_size + 1 - offset, maxInitialAlloc)
```

Comment:
That won't work. One reason for the seemingly awkward way we currently implemented read (with the loop below) is that in Unix there are files that report size 0 yet when you start reading from them, reading works. So the only way to read from those files is to consider `statbuf.st_size` of only advisory value, not ground truth.

The new code assumes `statbuf.st_size` as ground truth, which won't work.

Reply:
Do you specifically mean special files like /dev/zero or /dev/random? I've certainly never heard of normal files reporting an incorrect size, but that could certainly happen with special files. It had never occurred to me to use functions like /dev/zero or /dev/random with std.file, but I just tried it (with a value for `upTo`, of course), and it does work.

Reply:
Yah, it's the special files. People need to read those as well... there was a bug report about it involving a popular special file, forgot which that was.